Source/WebCore/Modules/cookie-store/CookieStore.cpp

```diff
@@ -210,79 +241,111 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+
+            // FIXME: <rdar://85515842> Obtain the encoded length without allocating and encoding.
+            if (cookie.domain.utf8().length() > maximumAttributeValueSize) {
+                promise->reject(Exception { TypeError, makeString("The size of the domain must not be greater than ", maximumAttributeValueSize, " bytes") });
+                return;
+            }
         }
 
-        // FIXME: <rdar://85515842> Obtain the encoded length without allocating and encoding.
-        if (cookie.domain.utf8().length() > maximumAttributeValueSize) {
-            promise->reject(Exception { TypeError, makeString("The size of the domain must not be greater than ", maximumAttributeValueSize, " bytes") });
-            return;
+        cookie.path = WTFMove(options.path);
+        if (!cookie.path.isNull()) {
+            if (!cookie.path.startsWith('/')) {
+                promise->reject(Exception { TypeError, "The path must begin with a '/'"_s });
+                return;
+            }
+
+            if (!cookie.path.endsWith('/'))
+                cookie.path = cookie.path + '/';
+
+            // FIXME: <rdar://85515842> Obtain the encoded length without allocating and encoding.
+            if (cookie.path.utf8().length() > maximumAttributeValueSize) {
+                promise->reject(Exception { TypeError, makeString("The size of the path must not be greater than ", maximumAttributeValueSize, " bytes") });
+                return;
+            }
         }
-    }
 
-    cookie.path = WTFMove(options.path);
-    if (!cookie.path.isNull()) {
-        if (!cookie.path.startsWith('/')) {
-            promise->reject(Exception { TypeError, "The path must begin with a '/'"_s });
-            return;
+        if (options.expires)
+            cookie.expires = *options.expires;
+
+        switch (options.sameSite) {
+        case CookieSameSite::Strict:
+            cookie.sameSite = Cookie::SameSitePolicy::Strict;
+            break;
+        case CookieSameSite::Lax:
+            cookie.sameSite = Cookie::SameSitePolicy::Lax;
+            break;
+        case CookieSameSite::None:
+            cookie.sameSite = Cookie::SameSitePolicy::None;
+            break;
         }
 
-        if (!cookie.path.endsWith('/'))
-            cookie.path = cookie.path + '/';
+        ALWAYS_LOG_WITH_STREAM(stream << "rupin log - point 4 - setting completion handler");
 
-        // FIXME: <rdar://85515842> Obtain the encoded length without allocating and encoding.
-        if (cookie.path.utf8().length() > maximumAttributeValueSize) {
-            promise->reject(Exception { TypeError, makeString("The size of the path must not be greater than ", maximumAttributeValueSize, " bytes") });
-            return;
-        }
-    }
+        auto& cookieJar = page->cookieJar();
+        auto completionHandler = [promise = WTFMove(promise)] (bool setSuccessfully) {
+            ALWAYS_LOG_WITH_STREAM(stream << "rupin log - point 7 - in completionHandler");
```

Comment:
I assume this is getting called on the main thread, even in the worker case, which means you're resolving the promise on the wrong thread even in the success case.